src/sync.cpp

```diff
@@ -166,10 +206,16 @@ static void pop_lock()
     }
 }
 
-void EnterCritical(const char* pszName, const char* pszFile, int nLine, void* cs, bool fTry)
+template <typename MutexType>
+void EnterCritical(const char* pszName, const char* pszFile, int nLine, MutexType* cs, bool fTry)
 {
     push_lock(cs, CLockLocation(pszName, pszFile, nLine, fTry, util::ThreadGetInternalName()));
 }
+template void EnterCritical(const char*, const char*, int, Mutex*, bool);
+template void EnterCritical(const char*, const char*, int, RecursiveMutex*, bool);
+template void EnterCritical(const char*, const char*, int, std::mutex*, bool);
+template void EnterCritical(const char*, const char*, int, std::recursive_mutex*, bool);
+template void EnterCritical(const char*, const char*, int, boost::mutex*, bool);
```

Comment:
no need to add dead code or code that is only used in tests

Reply:
These instantiations are required even if tests do not exist. Without them it wouldn't link with errors like:
```
ld: error: undefined symbol: void EnterCritical<boost::mutex>(char const*, char const*, int, boost::mutex*, bool)
>>> referenced by checkqueue.h:184 (src/checkqueue.h:184)
>>>               libbitcoin_server_a-validation.o:(CCheckQueueControl<CScriptCheck>::CCheckQueueControl(CCheckQueue<CScriptCheck>*)) in archive libbitcoin_server.a
```

Reply:
Yeah, this is solved in #18710 :)

Reply:
So when these two PRs meet and if that was the last usage of `boost::mutex`, then this line can be removed: `template void EnterCritical(const char*, const char*, int, boost::mutex*, bool);`.

Reply:
I tried compiling locally without it and it worked https://github.com/bitcoin/bitcoin/pull/19337#issuecomment-669211924

Reply:
Just for reference - the two PRs met in `master` and afterwards that line was removed in https://github.com/bitcoin/bitcoin/pull/21010.

Thanks, @fanquake!